catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,422 @@
+                          source.getId(),
+                          valueName);
+                      return source;
+                    }));
 
-    LOGGER.debug("Cancelling scheduled polling.");
+    // Remove orphaned SourceKeys or SourceKeys for destroyed Sources from the cache
+    final Set<SourceKey> sourceKeysInTheCache = cache.asMap().keySet();
+    final Set<SourceKey> currentSourceKeys = currentSourceKeyAndSourceMap.keySet();
+    final Set<SourceKey> sourceKeysToRemoveFromTheCache =
+        Sets.difference(sourceKeysInTheCache, currentSourceKeys);
+    if (LOGGER.isDebugEnabled() && !sourceKeysToRemoveFromTheCache.isEmpty()) {
+      LOGGER.debug(
+          "Found {} entries to remove from the cache: {}. This is caused by when sources are destroyed or modified in a way that causes their SourceKey to be different since the last poll",
+          sourceKeysToRemoveFromTheCache.size(),
+          StringUtils.join(sourceKeysToRemoveFromTheCache, ", "));
+      sourceKeysToRemoveFromTheCache
+          .stream()
+          .peek(sourceKey -> LOGGER.debug("Removing entry id={} from the cache", sourceKey.id))
+          .forEach(cache::invalidate);
+    }
 
-    runner.shutdown();
+    final int currentSourcesCount = currentSources.size();
 
-    handle.cancel(true);
+    // Loop through each of the currentSources
+    LOGGER.trace(
+        "Getting the {} for each of the current {} sources", valueName, currentSourcesCount);
+    final Map<Future, Pair<SourceKey, Optional<T>>> futureSourceKeyAndOptionalOldValueMap =
+        new HashMap<>();
+    for (final Map.Entry<SourceKey, Source> sourceKeyAndSourceEntry :
+        currentSourceKeyAndSourceMap.entrySet()) {
+      final Source source = sourceKeyAndSourceEntry.getValue();
+      final SourceKey sourceKey = sourceKeyAndSourceEntry.getKey();
 
-    scheduler.shutdownNow();
+      final Optional<T> optionalOldValue = Optional.ofNullable(cache.getIfPresent(sourceKey));
+      if (!optionalOldValue.isPresent()) {
+        LOGGER.debug(
+            "Found a new SourceKey {}. This is caused when a source is created or when a source is modified in a way that cause its SourceKey to be different since the last poll (or there were threading issues--see warning logs)",
+            sourceKey);
+      }
+
+      // Kick off a thread to find a new value
+      LOGGER.trace("Starting a process to get the {} for source {}", valueName, sourceKey);
+      final ListenableFuture getNewValueFuture =
+          Futures.withTimeout(
+              // TODO Handle RejectedExecutionException?
+              pollThreadPool.submit(
+                  () -> {
+                    try {
+                      final T newValue = getValueFunction.apply(source);
+
+                      // TODO Is there a simpler way to cache the value as soon as we know
+                      // it (e.g. handling futures by first completed)?
```

Comment:
❓

Reply:
This is left over from when I was working on it and is no longer relevant. 🙂